### PR TITLE
Handle user-defined types for populate struct

### PR DESCRIPTION
**Pull request by shawnburke**
I discovered that PopulateStruct fails in cases where there is:

1) A secondary provider that's serving values as strings (e.g. the env provider)
2) A user-defined type in a struct.

This handles this by falling back to comparing reflect.Kind rather than the type of the value to be set for string cases.

Added the YARPC test case that triggered this.

### CODE REVIEW

**Comment by sectioneight** [comment]
Alex recently re did the populate stuff to be much more flexible. Have you tried on the latest dev build?

**Comment by alsamylkin** [comment]
We do support extending these types if they implement TextUnmarshaller interface. In case of aliases we can try to convert these types as actual builtins first, and then fallback to TextUnmarshaller.

File: config/decoder.go [hunk 1]
```diff
@@ -119,13 +119,23 @@ func (d *decoder) getGlobalProvider() Provider {
...
 
 // Sets value to a primitive type.
-func (d *decoder) scalar(childKey string, value reflect.Value, def string) error {
+func (d *decoder) scalar(childKey string, value reflect.Value, def string) (err error) {
 	valueType := value.Type()
 	global := d.getGlobalProvider()
 	var val interface{}
 
+	defer func() {
+		if r := recover(); r != nil {
```
Comment (alsamylkin) [inline config/decoder.go]: Out of curiosity, what triggers panic? I've tried testcases and the only failing case was with port being 0, not 4324..

File: config/value.go [hunk 2]
```diff
@@ -304,16 +304,30 @@ func convertValue(value interface{}, targetType reflect.Type) (interface{}, erro
...
+		// 	return strconv.Atoi(v)
+		// case *bool:
+		// 	return strconv.ParseBool(v)
 		case *time.Duration:
 			return time.ParseDuration(v)
 		case encoding.TextUnmarshaler:
 			err := t.UnmarshalText([]byte(v))
 			// target should have a pointer receiver to be able to change itself based on text
 			return reflect.ValueOf(target).Elem().Interface(), err
+		default:
```
Comment (alsamylkin) [inline config/value.go]: This is going to work only if provider returns a string, not an e.g. int for a Port type, right?

Comment (shawnburke) [inline config/value.go]: Int works fine, the runtime will treat it like the user-defined types.  The provider does return a string, that's the testcase.

The catalyst providers are all string based.

File: config/decoder.go (hunk 1 above)
Comment (shawnburke) [inline config/decoder.go]: Did you try the test cases w/o the fixes?

```
--- FAIL: TestNestedStructs (0.00s)
panic: interface conversion: interface {} is string, not int [recovered]
	panic: interface conversion: interface {} is string, not int

goroutine 43 [running]:
testing.tRunner.func1(0xc42007f520)
	/usr/local/Cellar/go/1.8/libexec/src/testing/testing.go:622 +0x29d
panic(0x12f2760, 0xc420141140)
	/usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
go.uber.org/fx/config.convertValueFromStruct(0x12d90e0, 0xc420167060, 0x149b000, 0x12d88a0, 0x149b000, 0x12d88a0, 0x12d88a0, 0xc42016d030, 0x182, 0x12d90e0, ...)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:68 +0x16f
go.uber.org/fx/config.(*decoder).scalar(0xc4200f3ca0, 0xc420167008, 0x8, 0x12d88a0, 0xc42016d030, 0x182, 0x0, 0x0, 0x3, 0x0)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:198 +0xa89
go.uber.org/fx/config.(*decoder).unmarshal(0xc4200f3ca0, 0xc420167008, 0x8, 0x12d88a0, 0xc42016d030, 0x182, 0x0, 0x0, 0x10, 0xc420167000)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:409 +0x1ef
go.uber.org/fx/config.(*decoder).valueStruct(0xc4200f3ca0, 0x12bbf77, 0x4, 0x12c8f00, 0xc42016d020, 0x16, 0xc420166f80)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:379 +0x381
go.uber.org/fx/config.(*decoder).object(0xc4200f3ca0, 0x12bbf77, 0x4, 0x12c8f00, 0xc4200ad3a8, 0x196, 0x6, 0x12bbf77)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:345 +0x233
go.uber.org/fx/config.(*decoder).unmarshal(0xc4200f3ca0, 0x12bbf77, 0x4, 0x12c8f00, 0xc4200ad3a8, 0x196, 0x0, 0x0, 0x48, 0xc420166f20)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:411 +0x39c
go.uber.org/fx/config.(*decoder).valueStruct(0xc4200f3ca0, 0x0, 0x0, 0x12c90c0, 0xc4200ad360, 0x16, 0xc420166730)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:379 +0x381
go.uber.org/fx/config.(*decoder).object(0xc42004dca0, 0x0, 0x0, 0x12c90c0, 0xc4200ad360, 0x16, 0x1, 0x0)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:345 +0x233
go.uber.org/fx/config.(*decoder).unmarshal(0xc42004dca0, 0x0, 0x0, 0x131a160, 0xc4200ad360, 0x199, 0x0, 0x0, 0x0, 0xed05f8566)
	/Users/shawn/go/src/go.uber.org/fx/config/decoder.go:411 +0x39c
go.uber.org/fx/config.Value.PopulateStruct(0x1497360, 0xc42016c6f0, 0x1497480, 0xc420166730, 0x0, 0x0, 0x12ee200, 0xc42016c4b0, 0x1, 0x0, ...)
	/Users/shawn/go/src/go.uber.org/fx/config/value.go:341 +0x14d
go.uber.org/fx/config.TestNestedStructs(0xc42007f520)
	/Users/shawn/go/src/go.uber.org/fx/config/config_test.go:244 +0x322
testing.tRunner(0xc42007f520, 0x135a948)
	/usr/local/Cellar/go/1.8/libexec/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/local/Cellar/go/1.8/libexec/src/testing/testing.go:697 +0x2ca
exit status 2
FAIL	go.uber.org/fx/config	0.019s
```

**Comment by shawnburke** [comment]
@alsamylkin here is the error I get on the test cases rebased against dev. 

I don't understand your description, can you please take the test cases and make them work then? I'm happy to close this but we need this scenario working.

```
--- FAIL: TestRPCPortField (0.00s)
        Error Trace:    config_test.go:550
	Error:      	Received unexpected error:
	            	can't convert "4324" to integer type "config.Port"
FAIL
exit status 1
```

**Comment by alsamylkin** [comment]
Panics should be fixed by https://github.com/uber-go/fx/pull/415, because we were using type assertions to convert code, that's why you don't see them with latest changes.

Sure, I'll take it over: GFM-449.

**Pull request closed without merging**